pkg/alertdog/alertdog.go

```diff
@@ -0,0 +1,165 @@
+package alertdog
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/PagerDuty/go-pagerduty"
+	"github.com/prometheus/alertmanager/template"
+
+	"github.com/errm/alertdog/pkg/alertmanager"
+)
+
+type Alertmanager interface {
+	Alert(alertmanager.Alert) error
+	Resolve(alertmanager.Alert) error
+}
+
+type Pagerduty interface {
+	ManageEvent(event pagerduty.V2Event) (*pagerduty.V2EventResponse, error)
+}
+
+type Alertdog struct {
+	AlertmanagerEndpoints []string `yaml:"alertmanager_endpoints"`
+	Expected              []*Prometheus
+	CheckInterval         time.Duration
+	Expiry                time.Duration
+	PagerDutyKey          string
+	PagerDutyRunbookURL   string
+
+	mu           sync.RWMutex
+	checkedIn    time.Time
+	alertmanager Alertmanager
+	pagerduty    Pagerduty
+}
+
+func (a *Alertdog) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	defaultInterval, _ := time.ParseDuration("2m")
+	a.CheckInterval = defaultInterval
+	defaultExpiry, _ := time.ParseDuration("5m")
+	a.Expiry = defaultExpiry
+	type plain Alertdog
+	return unmarshal((*plain)(a))
+}
+
+func (a *Alertdog) Setup() {
+	a.alertmanager = alertmanager.Alertmanager{Endpoints: a.AlertmanagerEndpoints, Expiry: a.CheckInterval * 2}
+	a.pagerduty = PagerdutyClient{}
+}
+
+func (a *Alertdog) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	var data template.Data
+	if err := json.NewDecoder(r.Body).Decode(&data); err != nil {
+		log.Printf("Webhook body invalid, skipping request: %s", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	for _, alert := range data.Alerts {
+		a.processWatchdog(alert)
+	}
+	w.WriteHeader(http.StatusOK)
+	return
+}
+
+func (a *Alertdog) processWatchdog(alert template.Alert) {
+	a.CheckIn()
```

Comment:
Should we run this for each alert, or once for the http request?

Reply:
This is to just detect if alertmanager is up at all ... so just running it once for each request seems like the correct thing to do ...